// containers/phonePage/phonePage.js
import React, { Component } from 'react'
import { connect } from 'react-redux'
import * as R from 'ramda'
import { Link } from 'react-router-dom'

import { fetchPhoneById, addPhoneToBasket } from 'action'
import { getPhoneById } from 'selectors'
import BasketCart from 'components/basketCart/basketCart'
import Spinner from 'components/spinner/spinner'

class PhonePage extends Component {
  componentDidMount() {
    const id = this.props.match.params.id
    this.props.fetchPhoneById(id)
  }

  renderFields() {
    const { phone } = this.props
    const columnField = R.compose(
      R.toPairs,
      R.pick([
        'cpu',
        'camera',
        'size',
        'weight',
        'display',
        'battery',
        'memory'
      ])
    )(phone)

    return columnField.map(([key, value]) => (
      <div className='column' key={key}>
        <div className="ab-details-title">
          {key}
        </div>
        <div className="ab-details-info">
          {value}
        </div>
      </div>
    ))
  }

  renderFields_2() {
    const { phone } = this.props
    const filteredKeys = Object.keys(phone).filter(key => (
      key === 'cpu' ||
      key === 'camera' ||
      key === 'size' ||
      key === 'weight' ||
      key === 'display' ||
      key === 'battery' ||
      key === 'memory'
    ))

    return filteredKeys.map(key => (
      <div className='column' key={key}>
        <div className="ab-details-title">
          {key}
        </div>
        <div className="ab-details-info">
          {phone[key]}
        </div>
      </div>
    ))
  }

  renderContent() {
    const { phone } = this.props
    return (
      <div className='thumbnail'>
        <div className="row">

          <div className="col-md-6">
            <img
              src={phone.image}
              alt={phone.name}
              className="img-thumbnail"
            />
          </div>
          <div className="col-md-6">
            {this.renderFields()}
          </div>

        </div>
        <div className="caption-full">
          <h4 className="pull-right">
            ${phone.price}
          </h4>
          <h4>{phone.name}</h4>
          <p>{phone.description}</p>
        </div>
      </div>
    )
  }

  renderSidebar() {
    const { phone, addPhoneToBasket } = this.props
    return (
      <div>
        <p className="lead">Quick shop</p>
        <BasketCart />
        <div className="form-group">
          <h1>{phone.name}</h1>
          <h2>${phone.price}</h2>
          <Link
            to='/'
            className='btn btn-info btn-block'
          >
            Back to store
          </Link>
          <button
            className="btn btn-success btn-block"
            onClick={() => addPhoneToBasket(phone.id)}
          >
            Add to cart
          </button>
        </div>
      </div>
    )
  }

  render() {
    const { phone, loading } = this.props
    console.log('loading', loading)

    if (loading) {
      return <Spinner />
    }

    return (
      <div className='view-container'>
        <div className="container">
          <div className="row">

            <div className="col-md-9">
              {phone && this.renderContent()}
            </div>

            <div className="col-md-3">
              {phone && this.renderSidebar()}
            </div>

          </div>
        </div>
      </div>
    )
  }
}

const mapStateToProps = state => {
  return {
    phone: getPhoneById(state, state.phonePage.id),
    loading: state.phonePage.loading
  }
}

const mapDispatchToProps = {
  fetchPhoneById,
  addPhoneToBasket
}

export default connect(mapStateToProps, mapDispatchToProps)(PhonePage)
